### PR TITLE
osd/OSDMap: improve upmap calculation

### PR DESCRIPTION
**Pull request by liewegas**
The current version does not remap across hosts or racks (or other intervening
CRUSH bucket types)--it only moves PGs within the same lowest-level bucket referenced
by the rule.  This change fixes that.

On a test map (lab cluster) I'm able to get *all* devices to +/- 1 PG.

Also fix a bug or two.

### CODE REVIEW

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -3407,15 +3407,25 @@ int OSDMap::calc_pg_upmaps(
   CephContext *cct,
```
Comment (dachary) [inline src/osd/OSDMap.cc]: s/"equivlanet"/"equivalentt"/ in the commit message

File: src/crush/CrushWrapper.cc [hunk 2]
```diff
@@ -2287,6 +2315,50 @@ int CrushWrapper::_choose_type_stack(
...
+	      ldout(cct, 10) << " bucket " << o[pos] << " has no underfull targets and "
+			     << ">0 leaves " << leaves[pos] << " is overfull; alts "
+			     << underfull_buckets[j]
+			     << dendl;
+	      for (auto alt : underfull_buckets[j]) {
+		if (std::find(o.begin(), o.end(), alt) == o.end()) {
+		  // see if alt has the same parent
+		  if (j == 0 ||
+		      get_parent_of_type(o[pos], stack[j-1].first) ==
+		      get_parent_of_type(alt, stack[j-1].first)) {
```
Comment (dachary) [inline src/crush/CrushWrapper.cc]: border case if both get_parent_of_type return 0

**Comment by dachary** [comment]
@liewegas It looks like this could end up with a PG being mapped twice to the same host, via the remapping. Or am I missing part of the logic ? (maybe you replied already as I posted the same question before. I don't remember reading the response though).

**Comment by liewegas** [comment]
For each type in the "stack" we are building a vector of distinct items (just like crush would).  Those items still have to be unique.  In _choose_type_stack we try to swap out overfull items (leaves as before, now also buckets) for underfull items, but those swaps still check to make sure the items in the vector are unique.  So if the stack is something like [(host,3),(device,1)] we'll still get three unique hosts.. it just might be a different host (and device).

**Comment by dachary** [comment]
I missed something indeed, thanks for explaining.

File: src/crush/CrushWrapper.cc [hunk 3]
```diff
@@ -1218,6 +1229,11 @@ int CrushWrapper::get_rule_weight_osd_map(unsigned ruleno, map<int,float> *pmap)
   crush_rule *rule = crush->rules[ruleno];
 
   // build a weight map for each TAKE in the rule, and then merge them
+
+  // FIXME: if there are multiple takes that place a different number of
+  // objects we do not take that into account.  (Also, note that doing this
+  // right is also a function of the pool, since the crush rule
+  // might choose 2 + choose 2 but pool size may only be 3.)
```
Comment (dachary) [inline src/crush/CrushWrapper.cc]: side note: the border case of weights 5 1 1 1 1 should be handled in this function by reducing the weight 5 to 4

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-16_22:22:40-rados-wip-sage-testing2---basic-smithi/

**Pull request merged**